gnark_backend_ffi/main.go

```diff
@@ -103,28 +163,39 @@ func VerifyWithMeta(rawr1cs string, proof string) bool {
 //export VerifyWithVK
 func VerifyWithVK(rawr1cs string, proof string, verifyingKey string) bool {
 	// Create R1CS.
-	r1cs := cs_bls12381.NewR1CS(1)
+	r1cs := cs_bn254.NewR1CS(1)
 
 	// Add variables.
+	witness, err := witness.New(r1cs.CurveID().ScalarField())
+	if err != nil {
+		log.Fatal(err)
+	}
+	witness.Fill(0, 0, nil)
 
 	// Add constraints.
 
 	// Deserialize proof.
 	p := groth16.NewProof(r1cs.CurveID())
-	_, err_p := p.ReadFrom(bytes.NewReader([]byte(proof)))
-	if err_p != nil {
-		log.Fatal(err_p)
+	_, err = p.ReadFrom(bytes.NewReader([]byte(proof)))
+	if err != nil {
+		log.Fatal(err)
 	}
 
 	// Deserialize verifying key.
 	vk := groth16.NewVerifyingKey(r1cs.CurveID())
-	_, err_vk := vk.ReadFrom(bytes.NewReader([]byte(verifyingKey)))
-	if err_vk != nil {
-		log.Fatal(err_vk)
+	_, err = vk.ReadFrom(bytes.NewReader([]byte(verifyingKey)))
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	// Retrieve public inputs.
+	publicInputs, err := witness.Public()
+	if err != nil {
+		log.Fatal(err)
 	}
 
 	// Verify.
-	if groth16.Verify(p, vk, nil) != nil {
+	if groth16.Verify(p, vk, publicInputs) != nil {
 		return false
 	}
 
```

Comment:
Same here:
```go
return groth16.Verify(p, vk, publicInputs) == nil
```

Reply:
This was confusing for me at the time but it is because of how Go's error handling works. In general, if you want to return a value on a function that can fail, you'll return something like this: `result some_type, err error`. If `err` is `nil` this means that there was no error during the function execution. This is the same for this case, only that `groth16.Verify` returns `error` instead of `result bool, err error` meaning that it would return `nil` if there was no error during the execution.